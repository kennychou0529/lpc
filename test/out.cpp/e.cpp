
#include <lpcRuntime.h>

#line 0 "e.pas"

//================================================================================
// command line argument map
_Filename _FilenameMapEntries[] =
{
    { "input" },
    { "output" },
};

_Filename* _FilenameMap = _FilenameMapEntries;
const int _FilenameMapSize = _countof(_FilenameMapEntries);


#line 0
//================================================================================
// scope: global scope (level : 0)

// types
typedef char T_boolean;
typedef char T_char;
typedef int T_integer;
typedef double T_real;
typedef _T_Text T_text;

// constants
const T_boolean C_false = 0;
const T_integer C_maxint = 2147483647;
const T_boolean C_true = 1;

#line 2
//================================================================================
// scope: program scope (level : 1)

// types
typedef _T_Range<0, 255, unsigned __int8> T_digit;
typedef _T_Array< 0, 1007, T_digit > T_digitarray;
_PTR_TYPE(T_pointer_1, T_digitarray)

// forward declarations
void P_initinteger(void* _slink, T_digitarray& x, T_integer n);
void P_divide(void* _slink, T_digitarray& x, T_integer Xs, T_integer n, T_digitarray& y, T_integer& Ys);
void P_add(void* _slink, T_digitarray& s, T_digitarray& x, T_integer Xs);
void P_sub(void* _slink, T_digitarray& s, T_digitarray& x, T_integer Xs);

// constants
const T_integer C_NDIGITS = 1007;
const T_integer C_NPRINT = 1000;

// program variables
T_integer i;
T_text Input;
T_text Output;
T_pointer_1 s;
T_pointer_1 t;
T_integer Ts;
T_pointer_1 x;
T_integer Xs;

// function body
void P_()
{
    
    // initializers
    Input.init(0);
    Output.init(1);
    
    // subroutine body
    #line 102
    _new(s._ptr);
    
    #line 102
    _new(x._ptr);
    
    #line 103
    P_initinteger(nullptr, (*s._ptr), 0);
    
    #line 104
    P_initinteger(nullptr, (*x._ptr), 1);
    
    #line 105
    Xs = 0;
    
    #line 106
    P_add(nullptr, (*s._ptr), (*x._ptr), Xs);
    
    #line 107
    i = 0;
    
    #line 108
    do
    {
        #line 109
        i = (i + 1);
        
        #line 110
        P_divide(nullptr, (*x._ptr), Xs, i, (*x._ptr), Xs);
        
        #line 111
        P_add(nullptr, (*s._ptr), (*x._ptr), Xs);
        
        #line 112
        Output << "Series: " << _format(((double)(100 * Xs) / (C_NDIGITS + 1)), 5, 2) << '%' << _WRITELN;
        
        { /* NOP */ }
    }
    while(!(Xs > C_NDIGITS));
    
    #line 114
    Output << _WRITELN;
    
    #line 115
    Output << _format(' ', 45, 0) << "e = " << _format((*s._ptr)[0], 1, 0) << '.' << _WRITELN;
    
    #line 116
    i = 0;
    
    #line 117
    _FOR_TO(i, 1, C_NPRINT)
    {
        #line 118
        Output << _format((*s._ptr)[i], 1, 0);
        
        #line 119
        if((_mod(i , 1000) == 0))
        {
            #line 119
            Output << _WRITELN;
        }
        
        #line 120
        if((_mod(i , 100) == 0))
        {
            #line 120
            Output << _WRITELN;
        }
        else
        {
            #line 121
            if((_mod(i , 10) == 0))
            {
                #line 121
                Output << ' ';
            }
        }
        
        { /* NOP */ }
    }
    _FOR_END(i)
    
    #line 123
    Output << _WRITELN;
    
    #line 124
    Output << "Final digits: ";
    
    #line 125
    _FOR_TO(i, (C_NPRINT + 1), C_NDIGITS)
    {
        #line 126
        Output << _format((*s._ptr)[i], 1, 0);
        
        { /* NOP */ }
    }
    _FOR_END(i)
    
    #line 128
    Output << _WRITELN;
    
    { /* NOP */ }
}


#line 17
//================================================================================
// scope: initinteger (level : 2)

// activation record
struct Frame_initinteger
{
    // parameters
    T_digitarray& x;
    T_integer n;

    // locals
    T_integer i;
};

// function body
void P_initinteger(void* _slink, T_digitarray& x, T_integer n)
{
    // prologue/frame definition
    Frame_initinteger _F = {x, n };
    
    assert(nullptr == _slink);
    
    // subroutine body
    #line 21
    _F.x[0] = _F.n;
    
    #line 22
    _FOR_TO(_F.i, 1, C_NDIGITS)
    {
        #line 22
        _F.x[_F.i] = 0;
    }
    _FOR_END(_F.i)
    
    { /* NOP */ }
}


#line 25
//================================================================================
// scope: divide (level : 2)

// activation record
struct Frame_divide
{
    // parameters
    T_digitarray& x;
    T_integer Xs;
    T_integer n;
    T_digitarray& y;
    T_integer& Ys;

    // locals
    T_integer c;
    T_integer i;
};

// function body
void P_divide(void* _slink, T_digitarray& x, T_integer Xs, T_integer n, T_digitarray& y, T_integer& Ys)
{
    // prologue/frame definition
    Frame_divide _F = {x, Xs, n, y, Ys };
    
    assert(nullptr == _slink);
    
    // subroutine body
    #line 31
    _F.c = 0;
    
    #line 32
    _FOR_TO(_F.i, _F.Xs, C_NDIGITS)
    {
        #line 33
        _F.c = ((10 * _F.c) + _F.x[_F.i]);
        
        #line 34
        _F.y[_F.i] = (_F.c / _F.n);
        
        #line 35
        _F.c = _mod(_F.c , _F.n);
        
        { /* NOP */ }
    }
    _FOR_END(_F.i)
    
    #line 37
    _F.Ys = _F.Xs;
    
    #line 38
    while(((_F.Ys <= C_NDIGITS) && (_F.y[_F.Ys] == 0)))
    {
        #line 38
        _F.Ys = (_F.Ys + 1);
    }
    
    { /* NOP */ }
}


#line 41
//================================================================================
// scope: add (level : 2)

// activation record
struct Frame_add
{
    // parameters
    T_digitarray& s;
    T_digitarray& x;
    T_integer Xs;

    // locals
    T_integer c;
    T_integer i;
};

// function body
void P_add(void* _slink, T_digitarray& s, T_digitarray& x, T_integer Xs)
{
    // prologue/frame definition
    Frame_add _F = {s, x, Xs };
    
    assert(nullptr == _slink);
    
    // subroutine body
    #line 46
    _F.c = 0;
    
    #line 47
    _FOR_DOWNTO(_F.i, C_NDIGITS, _F.Xs)
    {
        #line 48
        _F.c = ((_F.s[_F.i] + _F.x[_F.i]) + _F.c);
        
        #line 49
        if((_F.c >= 10))
        {
            #line 50
            _F.s[_F.i] = (_F.c - 10);
            
            #line 51
            _F.c = 1;
            
            { /* NOP */ }
        }
        else
        {
            #line 53
            _F.s[_F.i] = _F.c;
            
            #line 54
            _F.c = 0;
            
            { /* NOP */ }
        }
        
        { /* NOP */ }
    }
    _FOR_END(_F.i)
    
    #line 57
    _F.i = _F.Xs;
    
    #line 58
    while((_F.c != 0))
    {
        #line 59
        _F.i = (_F.i - 1);
        
        #line 60
        _F.c = (_F.s[_F.i] + _F.c);
        
        #line 61
        if((_F.c >= 10))
        {
            #line 62
            _F.s[_F.i] = (_F.c - 10);
            
            #line 63
            _F.c = 1;
            
            { /* NOP */ }
        }
        else
        {
            #line 65
            _F.s[_F.i] = _F.c;
            
            #line 66
            _F.c = 0;
            
            { /* NOP */ }
        }
        
        { /* NOP */ }
    }
    
    { /* NOP */ }
}


#line 71
//================================================================================
// scope: sub (level : 2)

// activation record
struct Frame_sub
{
    // parameters
    T_digitarray& s;
    T_digitarray& x;
    T_integer Xs;

    // locals
    T_integer c;
    T_integer i;
};

// function body
void P_sub(void* _slink, T_digitarray& s, T_digitarray& x, T_integer Xs)
{
    // prologue/frame definition
    Frame_sub _F = {s, x, Xs };
    
    assert(nullptr == _slink);
    
    // subroutine body
    #line 76
    _F.c = 0;
    
    #line 77
    _FOR_DOWNTO(_F.i, C_NDIGITS, _F.Xs)
    {
        #line 78
        _F.c = ((_F.s[_F.i] - _F.x[_F.i]) + _F.c);
        
        #line 79
        if((_F.c < 0))
        {
            #line 80
            _F.s[_F.i] = (_F.c + 10);
            
            #line 81
            _F.c = (-1);
            
            { /* NOP */ }
        }
        else
        {
            #line 83
            _F.s[_F.i] = _F.c;
            
            #line 84
            _F.c = 0;
            
            { /* NOP */ }
        }
        
        { /* NOP */ }
    }
    _FOR_END(_F.i)
    
    #line 87
    _F.i = _F.Xs;
    
    #line 88
    while((_F.c != 0))
    {
        #line 89
        _F.i = (_F.i - 1);
        
        #line 90
        _F.c = (_F.s[_F.i] + _F.c);
        
        #line 91
        if((_F.c < 0))
        {
            #line 92
            _F.s[_F.i] = (_F.c + 10);
            
            #line 93
            _F.c = (-1);
            
            { /* NOP */ }
        }
        else
        {
            #line 95
            _F.s[_F.i] = _F.c;
            
            #line 96
            _F.c = 0;
            
            { /* NOP */ }
        }
        
        { /* NOP */ }
    }
    
    { /* NOP */ }
}

