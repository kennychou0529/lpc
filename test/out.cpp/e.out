Series:  0.00%
Series:  0.10%
Series:  0.10%
Series:  0.20%
Series:  0.30%
Series:  0.30%
Series:  0.40%
Series:  0.50%
Series:  0.60%
Series:  0.69%
Series:  0.79%
Series:  0.89%
Series:  0.99%
Series:  1.09%
Series:  1.29%
Series:  1.39%
Series:  1.49%
Series:  1.59%
Series:  1.79%
Series:  1.88%
Series:  1.98%
Series:  2.18%
Series:  2.28%
Series:  2.38%
Series:  2.58%
Series:  2.68%
Series:  2.88%
Series:  2.98%
Series:  3.08%
Series:  3.27%
Series:  3.37%
Series:  3.57%
Series:  3.67%
Series:  3.87%
Series:  4.07%
Series:  4.17%
Series:  4.37%
Series:  4.46%
Series:  4.66%
Series:  4.76%
Series:  4.96%
Series:  5.16%
Series:  5.26%
Series:  5.46%
Series:  5.65%
Series:  5.75%
Series:  5.95%
Series:  6.15%
Series:  6.25%
Series:  6.45%
Series:  6.65%
Series:  6.75%
Series:  6.94%
Series:  7.14%
Series:  7.34%
Series:  7.44%
Series:  7.64%
Series:  7.84%
Series:  8.04%
Series:  8.13%
Series:  8.33%
Series:  8.53%
Series:  8.73%
Series:  8.93%
Series:  9.03%
Series:  9.23%
Series:  9.42%
Series:  9.62%
Series:  9.82%
Series: 10.02%
Series: 10.12%
Series: 10.32%
Series: 10.52%
Series: 10.71%
Series: 10.91%
Series: 11.11%
Series: 11.31%
Series: 11.51%
Series: 11.61%
Series: 11.81%
Series: 12.00%
Series: 12.20%
Series: 12.40%
Series: 12.60%
Series: 12.80%
Series: 13.00%
Series: 13.19%
Series: 13.39%
Series: 13.59%
Series: 13.79%
Series: 13.99%
Series: 14.19%
Series: 14.38%
Series: 14.58%
Series: 14.78%
Series: 14.88%
Series: 15.08%
Series: 15.28%
Series: 15.48%
Series: 15.67%
Series: 15.87%
Series: 16.07%
Series: 16.27%
Series: 16.57%
Series: 16.77%
Series: 16.96%
Series: 17.16%
Series: 17.36%
Series: 17.56%
Series: 17.76%
Series: 17.96%
Series: 18.15%
Series: 18.35%
Series: 18.55%
Series: 18.75%
Series: 18.95%
Series: 19.15%
Series: 19.35%
Series: 19.54%
Series: 19.74%
Series: 19.94%
Series: 20.14%
Series: 20.44%
Series: 20.63%
Series: 20.83%
Series: 21.03%
Series: 21.23%
Series: 21.43%
Series: 21.63%
Series: 21.83%
Series: 22.02%
Series: 22.32%
Series: 22.52%
Series: 22.72%
Series: 22.92%
Series: 23.12%
Series: 23.31%
Series: 23.51%
Series: 23.71%
Series: 24.01%
Series: 24.21%
Series: 24.40%
Series: 24.60%
Series: 24.80%
Series: 25.00%
Series: 25.30%
Series: 25.50%
Series: 25.69%
Series: 25.89%
Series: 26.09%
Series: 26.29%
Series: 26.59%
Series: 26.79%
Series: 26.98%
Series: 27.18%
Series: 27.38%
Series: 27.68%
Series: 27.88%
Series: 28.08%
Series: 28.27%
Series: 28.47%
Series: 28.77%
Series: 28.97%
Series: 29.17%
Series: 29.37%
Series: 29.56%
Series: 29.86%
Series: 30.06%
Series: 30.26%
Series: 30.46%
Series: 30.75%
Series: 30.95%
Series: 31.15%
Series: 31.35%
Series: 31.65%
Series: 31.85%
Series: 32.04%
Series: 32.24%
Series: 32.54%
Series: 32.74%
Series: 32.94%
Series: 33.13%
Series: 33.43%
Series: 33.63%
Series: 33.83%
Series: 34.03%
Series: 34.33%
Series: 34.52%
Series: 34.72%
Series: 34.92%
Series: 35.22%
Series: 35.42%
Series: 35.62%
Series: 35.91%
Series: 36.11%
Series: 36.31%
Series: 36.61%
Series: 36.81%
Series: 37.00%
Series: 37.20%
Series: 37.50%
Series: 37.70%
Series: 37.90%
Series: 38.19%
Series: 38.39%
Series: 38.59%
Series: 38.89%
Series: 39.09%
Series: 39.29%
Series: 39.58%
Series: 39.78%
Series: 39.98%
Series: 40.28%
Series: 40.48%
Series: 40.67%
Series: 40.97%
Series: 41.17%
Series: 41.37%
Series: 41.67%
Series: 41.87%
Series: 42.06%
Series: 42.36%
Series: 42.56%
Series: 42.76%
Series: 43.06%
Series: 43.25%
Series: 43.45%
Series: 43.75%
Series: 43.95%
Series: 44.15%
Series: 44.44%
Series: 44.64%
Series: 44.84%
Series: 45.14%
Series: 45.34%
Series: 45.63%
Series: 45.83%
Series: 46.03%
Series: 46.33%
Series: 46.53%
Series: 46.73%
Series: 47.02%
Series: 47.22%
Series: 47.52%
Series: 47.72%
Series: 47.92%
Series: 48.21%
Series: 48.41%
Series: 48.71%
Series: 48.91%
Series: 49.11%
Series: 49.40%
Series: 49.60%
Series: 49.90%
Series: 50.10%
Series: 50.30%
Series: 50.60%
Series: 50.79%
Series: 51.09%
Series: 51.29%
Series: 51.49%
Series: 51.79%
Series: 51.98%
Series: 52.28%
Series: 52.48%
Series: 52.78%
Series: 52.98%
Series: 53.17%
Series: 53.47%
Series: 53.67%
Series: 53.97%
Series: 54.17%
Series: 54.46%
Series: 54.66%
Series: 54.96%
Series: 55.16%
Series: 55.36%
Series: 55.65%
Series: 55.85%
Series: 56.15%
Series: 56.35%
Series: 56.65%
Series: 56.85%
Series: 57.14%
Series: 57.34%
Series: 57.54%
Series: 57.84%
Series: 58.04%
Series: 58.33%
Series: 58.53%
Series: 58.83%
Series: 59.03%
Series: 59.33%
Series: 59.52%
Series: 59.82%
Series: 60.02%
Series: 60.32%
Series: 60.52%
Series: 60.81%
Series: 61.01%
Series: 61.21%
Series: 61.51%
Series: 61.71%
Series: 62.00%
Series: 62.20%
Series: 62.50%
Series: 62.70%
Series: 63.00%
Series: 63.19%
Series: 63.49%
Series: 63.69%
Series: 63.99%
Series: 64.19%
Series: 64.48%
Series: 64.68%
Series: 64.98%
Series: 65.18%
Series: 65.48%
Series: 65.67%
Series: 65.97%
Series: 66.17%
Series: 66.47%
Series: 66.67%
Series: 66.96%
Series: 67.16%
Series: 67.46%
Series: 67.66%
Series: 67.96%
Series: 68.15%
Series: 68.45%
Series: 68.65%
Series: 68.95%
Series: 69.25%
Series: 69.44%
Series: 69.74%
Series: 69.94%
Series: 70.24%
Series: 70.44%
Series: 70.73%
Series: 70.93%
Series: 71.23%
Series: 71.43%
Series: 71.73%
Series: 71.92%
Series: 72.22%
Series: 72.42%
Series: 72.72%
Series: 73.02%
Series: 73.21%
Series: 73.51%
Series: 73.71%
Series: 74.01%
Series: 74.21%
Series: 74.50%
Series: 74.70%
Series: 75.00%
Series: 75.20%
Series: 75.50%
Series: 75.79%
Series: 75.99%
Series: 76.29%
Series: 76.49%
Series: 76.79%
Series: 76.98%
Series: 77.28%
Series: 77.48%
Series: 77.78%
Series: 78.08%
Series: 78.27%
Series: 78.57%
Series: 78.77%
Series: 79.07%
Series: 79.27%
Series: 79.56%
Series: 79.86%
Series: 80.06%
Series: 80.36%
Series: 80.56%
Series: 80.85%
Series: 81.05%
Series: 81.35%
Series: 81.65%
Series: 81.85%
Series: 82.14%
Series: 82.34%
Series: 82.64%
Series: 82.94%
Series: 83.13%
Series: 83.43%
Series: 83.63%
Series: 83.93%
Series: 84.23%
Series: 84.42%
Series: 84.72%
Series: 84.92%
Series: 85.22%
Series: 85.52%
Series: 85.71%
Series: 86.01%
Series: 86.21%
Series: 86.51%
Series: 86.81%
Series: 87.00%
Series: 87.30%
Series: 87.50%
Series: 87.80%
Series: 88.10%
Series: 88.29%
Series: 88.59%
Series: 88.79%
Series: 89.09%
Series: 89.38%
Series: 89.58%
Series: 89.88%
Series: 90.08%
Series: 90.38%
Series: 90.67%
Series: 90.87%
Series: 91.17%
Series: 91.47%
Series: 91.67%
Series: 91.96%
Series: 92.16%
Series: 92.46%
Series: 92.76%
Series: 92.96%
Series: 93.25%
Series: 93.55%
Series: 93.75%
Series: 94.05%
Series: 94.25%
Series: 94.54%
Series: 94.84%
Series: 95.04%
Series: 95.34%
Series: 95.63%
Series: 95.83%
Series: 96.13%
Series: 96.43%
Series: 96.63%
Series: 96.92%
Series: 97.22%
Series: 97.42%
Series: 97.72%
Series: 97.92%
Series: 98.21%
Series: 98.51%
Series: 98.71%
Series: 99.01%
Series: 99.31%
Series: 99.50%
Series: 99.80%
Series: 100.00%

                                             e = 2.
7182818284 5904523536 0287471352 6624977572 4709369995 9574966967 6277240766 3035354759 4571382178 5251664274
2746639193 2003059921 8174135966 2904357290 0334295260 5956307381 3232862794 3490763233 8298807531 9525101901
1573834187 9307021540 8914993488 4167509244 7614606680 8226480016 8477411853 7423454424 3710753907 7744992069
5517027618 3860626133 1384583000 7520449338 2656029760 6737113200 7093287091 2744374704 7230696977 2093101416
9283681902 5515108657 4637721112 5238978442 5056953696 7707854499 6996794686 4454905987 9316368892 3009879312
7736178215 4249992295 7635148220 8269895193 6680331825 2886939849 6465105820 9392398294 8879332036 2509443117
3012381970 6841614039 7019837679 3206832823 7646480429 5311802328 7825098194 5581530175 6717361332 0698112509
9618188159 3041690351 5988885193 4580727386 6738589422 8792284998 9208680582 5749279610 4841984443 6346324496
8487560233 6248270419 7862320900 2160990235 3043699418 4914631409 3431738143 6405462531 5209618369 0888707016
7683964243 7814059271 4563549061 3031072085 1038375051 0115747704 1718986106 8739696552 1267154688 9570350354


Final digits: 0212109
